### ISSUE TITLE
"Check for SDK updates" option to check for latest Flutter bundled

### ISSUE THREAD

**Issue by FMorschel** [Dart-Code/Dart-Code#5127]
**Is your feature request related to a problem? Please describe.**
Flutter 3.22.1 at this current date, is using Dart 3.4.1 while Dart 3.4.2 is already released.

Every project I open I receive this warning:
![image](https://github.com/Dart-Code/Dart-Code/assets/52160996/fd58294c-9521-4ad2-8f88-adcee06fed81)

**Describe the solution you'd like**
I'd like the following option to stop being a boolean value and having an array of options:
![image](https://github.com/Dart-Code/Dart-Code/assets/52160996/01a31786-50aa-4f9a-90c7-1816a47445df)

Options like:
- Always
- Latest Flutter bundled
- Never

**Describe alternatives you've considered**
Waiting for a new Flutter release and hope that it syncs again while the Dart SDK doesn't release another update.

**Comment by FMorschel** [comment]
Since I created this and #5126 at the same time, maybe you missed this as well. Sorry for the ping, I was just not sure you ever saw this, @DanTup.

**Comment by DanTup** [comment]
Thanks - I had not seen this. I'll review the issue list, seems I didn't get email notifications for some of these.

I raised this in https://github.com/Dart-Code/Dart-Code/issues/4976 but was unable to reproduce it, and I couldn't find a code path for it (we are never supposed to prompt about Dart SDK versions when the Dart SDK is from Flutter).

Can you still reproduce this? What does the language status (click the `{}` next to `Dart` in the Status bar) show?

![image](https://github.com/Dart-Code/Dart-Code/assets/1078012/0eed36b8-969b-49c5-97f8-58ec4956de99)

And is the project you have open a Flutter or Dart project? (You can still use the Dart SDK from the Flutter SDK for a Dart project, and also still should not prompt).

**Comment by FMorschel** [comment]
Either a Dart-only project or a Flutter project both still show the message for the latest stable Flutter:
![image](https://github.com/Dart-Code/Dart-Code/assets/52160996/0650d3c9-e3fb-4e93-883a-030d8b4c0e30)

![image](https://github.com/Dart-Code/Dart-Code/assets/52160996/cc47880a-4daf-41b0-94e7-1b96c186fdca)

Not sure but if you can't reproduce it maybe it has something to do with the current SDK not being in my path (using `puro`)

`flutter doctor`
```console
[!] Flutter (Channel stable, 3.22.2, on Microsoft Windows [Version 10.0.22631.3737], locale pt-BR)
    ! Warning: `dart` on your path resolves to C:\Users\felip_0vh5fa6\.puro\shared\caches\dd37cefd4a94c91032a2e99bf4ca83c81425ed9d\dart-sdk\bin\dart.exe, which is not inside your current Flutter SDK
      checkout at C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter. Consider adding C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter\bin to the front of your path.
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[√] Chrome - develop for the web
[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.4.5)
[√] Android Studio (version 2024.1)
[√] VS Code (version 1.90.1)
[√] VS Code (version 1.88.0-insider)
[√] Connected device (4 available)
[√] Network resources

! Doctor found issues in 1 category.
```

**Comment by DanTup** [comment]
> Not sure but if you can't reproduce it maybe it has something to do with the current SDK not being in my path (using `puro`)

> ```
> ! Warning: `dart` on your path resolves to C:\Users\felip_0vh5fa6\.puro\shared\caches\dd37cefd4a94c91032a2e99bf4ca83c81425ed9d\dart-sdk\bin\dart.exe, which is not inside your current Flutter SDK
>      checkout at C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter. Consider adding C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter\bin to the front of your path.
>

Generally the extension doesn't care what's on your `PATH`. Besides locating an initial SDK, we shouldn't use it (because we allow overriding it and using anything from `PATH` would make things inconsistent). We should be finding your Flutter SDK first, and then taking the Dart SDK from `bin\cache\dart-sdk`.

Can you confirm there is a normal Dart SDK at `C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter\bin\cache\dart-sdk`?

If you open VS Code (and a project where this happens), then immediately run the **Dart: Open Extension Log** command, it'll open a log file that should detail the search for SDKs that might help. Are you able to share that (be sure to remove anything sensitive)?

**Comment by FMorschel** [comment]
Here it is:

[dart-code-startup-log-786b.txt](https://github.com/user-attachments/files/15915491/dart-code-startup-log-786b.txt)

**Comment by DanTup** [comment]
Thanks! I think this is the problem:

Following symlink: C:\Users\user\.puro\envs\stable\flutter\bin\cache\dart-sdk\bin\dart.exe -> C:\Users\user\.puro\shared\caches\edd8546116457bdf1c5bdfb13ecb9463d2bb5ed4\dart-sdk\bin\dart.exe

Your Flutter SDK has a symlink at `C:\Users\user\.puro\envs\stable\flutter\bin\cache\dart-sdk\bin\dart.exe` that points to another location. This is not a configuration we support - we assume the Flutter SDK is complete as Flutter would create it. In this case, we're using a Dart SDK that is not inside the Flutter SDK, and therefore treat it as a standalone SDK.

I'm not very familiar with Puro - do you know why it's symlinking this or whether you can control it?

**Comment by FMorschel** [comment]
In my company we have, as of right now, 5 different apps in production and 1 in development. We have set a different puro environment for each. 

https://github.com/flutter/flutter/issues/148692 is an example of cases we'd like to avoid.

While some of our projects were not implementing `TextEditingController`s, others were. The ones that would break we stationed until there were a fix and the others we upgraded to the latest stable.

**Comment by DanTup** [comment]
> We have set a different puro environment for each.

I'm not familiar with Puro so it's not obvious to me how this differs to setting the SDK in VS Code. For example if you had your SDKs at `~/flutter/(version)` and committed the `dart.flutterSdkPath` into each projects `.vscode/settings.json`, what would you be missing?

I guess you'd have to set the SDKs up manually whereas I presume Puro does that for you - but I wonder if it does anything else? There are a number of version mangers out there and they all have their own slightly-different quirks and often lead to issues (like this one) so it's useful for me to understand the reasons people use them (because it might impact how things work here).

Thanks!

**Comment by FMorschel** [comment]
Yes, exactly it. It downloads and manages them for us. When running `puro use environment` it replaces those values you mentioned for both vsCode and IntelliJ. It's also a _little_ bit easier to manage them when you can run, say:

```console
pruo upgrade env1 stable
puro upgrade env2 3.22.1
puro upgrade edge master
puro upgrade beta beta
```

That would be like going to those specific environment folders you suggested and upgrading them.

At least in my company this is all it does for us at the moment. There may be some other current use cases if you take a look at some closed issues or potentially new cases in open ones.

One thing I've not yet tried to do was to benchmark the download/install time for puro or flutter alone, but [here at their website](https://puro.dev/#performance) they talk about it being faster.

**Comment by DanTup** [comment]
Gotcha, thanks for the insight.

I've pushed a fix so we'll check both the original path and the resolved path to see if it's a Flutter SDK. I couldn't get Puro working (I filed an issue in their GH for this) but I created a symlink manually and could repro the issue before my fix, but not without.

I can't promise there won't be other issues with this kind of setup, but I assume if there were you'd have seen them by now.
